Objects/unicodeobject.c

```diff
@@ -14617,6 +14617,136 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
+        return NULL;
+    }
+    Py_ssize_t sz;
+    const char *str = PyUnicode_AsUTF8AndSize(obj, &sz);
+    if (str == NULL) {
+        return NULL;
+    }
+    if (strlen(str) != (size_t)sz) {
+        PyErr_SetString(PyExc_ValueError, "embedded null character");
+        return NULL;
+    }
+    return str;
+}
+
+static PyObject *
+fallback_to_tp_call(PyObject *type, Py_ssize_t nargs, Py_ssize_t nkwargs,
+                    PyObject *const *args, PyObject *kwnames)
+{
+    PyObject *tuple = PyTuple_New(nargs);
+    if (tuple == NULL) {
+        return NULL;
+    }
+    for (Py_ssize_t i = 0; i < nargs; i++) {
+        PyObject *value = Py_NewRef(args[i]);
+        PyTuple_SET_ITEM(tuple, i, value);
+    }
+    PyObject *dict = PyDict_New();
+    if (dict == NULL) {
+        Py_DECREF(tuple);
+        return NULL;
+    }
+    for (Py_ssize_t j = 0; j < nkwargs; j++) {
+        PyObject *key = PyTuple_GET_ITEM(kwnames, j);
+        PyObject *value = args[nargs + j];
+        if (PyDict_SetItem(dict, key, value) < 0) {
+            Py_DECREF(tuple);
+            Py_DECREF(dict);
+            return NULL;
+        }
+    }
+    return unicode_new(_PyType_CAST(type), tuple, dict);
+}
+
+static PyObject *
+unicode_vectorcall(PyObject *type, PyObject *const *args,
+                   size_t nargsf, PyObject *kwnames)
+{
+    assert(Py_Is(_PyType_CAST(type), &PyUnicode_Type));
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    Py_ssize_t nkwargs = (kwnames) ? PyTuple_GET_SIZE(kwnames) : 0;
+    if (nargs == 0 && nkwargs == 0) {
+        return unicode_get_empty();
+    }
+    PyObject *object = args[0];
+    if (nargs == 1 && nkwargs == 0) {
+        return PyObject_Str(object);
+    }
+    if (nargs + nkwargs == 2) {
+        const char *encoding;
+        const char *errors;
+        if (nkwargs == 1) {
+            PyObject *kw0 = PyTuple_GET_ITEM(kwnames, 0);
+            if (_PyUnicode_EqualToASCIIString(kw0, "encoding")) {
+                encoding = as_const_char(args[1], "encoding");
+                if (encoding == NULL) {
+                    return NULL;
+                }
+                errors = NULL;
+            }
+            else if (_PyUnicode_EqualToASCIIString(kw0, "errors")) {
+                errors = as_const_char(args[1], "errors");
+                if (errors == NULL) {
+                    return NULL;
+                }
+                encoding = NULL;
+            }
+            else {
+                PyErr_Format(PyExc_TypeError,
+                    "'%S' is an invalid keyword argument for str()", kw0);
+                return NULL;
+            }
+        }
+        else if (nkwargs == 0) {
+            encoding = as_const_char(args[1], "encoding");
+            errors = NULL;
+        }
+        else {
+            return fallback_to_tp_call(type, nargs, nkwargs, args, kwnames);
+        }
+        return PyUnicode_FromEncodedObject(object, encoding, errors);
+    }
+    if (nargs + nkwargs == 3) {
```

Comment:
I'm not sure that it's worth it to optimize the 3 arguments case using vectorcall.

Reply:
Yes, this is the real question: which cases do we want to optimise? It would be nice if we had some usage stats. Perhaps the Faster CPython team has stats. cc. @mdboom

Reply:
Using a regular expression, I only found 5 lines in the whole stdlib (excluding tests) which call str() with 3 arguments, and all these matching lines only use positional arguments:

```py
$ git grep '\<str *([^,()]\+,[^,()]\+,'
Lib/email/utils.py:        return str(rawbytes, charset, errors)
Lib/encodings/punycode.py:            base = str(text[:pos], "ascii", errors)
Lib/pickletools.py:        return str(data, 'utf-8', 'surrogatepass')
Lib/pickletools.py:        return str(data, 'utf-8', 'surrogatepass')
Lib/pickletools.py:        return str(data, 'utf-8', 'surrogatepass')
```

I modified `git grep` output to ignore tests, documentation and lines which care not str() calls (like Argument Clinic code).

Reply:
Well, optimising only for the positional-only argument cases would definitely make the PR smaller and the resulting code more maintainable.

Reply:
A PR for positional-only cases is up for comparison:

- #117746

Reply:
> There are only results in the documentation [...]

Well, people often copy the examples from the documentation, so I would expect the examples shown in the docs to be very common in the wild.